### PR TITLE
Add a basic numpy type

### PR DESCRIPTION
**Pull request by seanmcl**
Adds a Numpy representation and parser for .npy files. Has worked on different ndarrays I've saved manually. I'd like to add some tests but figure it would be nice to get some early feedback.

Sample output for the ndarray `[[2,3,4], [2,3,4]]`

```
$ .lake/build/bin/tensorlib parse-npy 1.npy
Parsing 1.npy...
{ header := { major := 1, minor := 0, descr := TensorLib.NumpyDtype.int64, fortranOrder := false, shape := [2, 3] },
  data := ByteArray of size 176,
  startIndex := 128,
  endIndex := 176 }
```

Hex dump of 1.py looks like so

```
00000000  93 4e 55 4d 50 59 01 00  76 00 7b 27 64 65 73 63  |.NUMPY..v.{'desc|
00000010  72 27 3a 20 27 3c 69 38  27 2c 20 27 66 6f 72 74  |r': '<i8', 'fort|
00000020  72 61 6e 5f 6f 72 64 65  72 27 3a 20 46 61 6c 73  |ran_order': Fals|
00000030  65 2c 20 27 73 68 61 70  65 27 3a 20 28 32 2c 20  |e, 'shape': (2, |
00000040  33 29 2c 20 7d 20 20 20  20 20 20 20 20 20 20 20  |3), }           |
00000050  20 20 20 20 20 20 20 20  20 20 20 20 20 20 20 20  |                |
00000060  20 20 20 20 20 20 20 20  20 20 20 20 20 20 20 20  |                |
00000070  20 20 20 20 20 20 20 20  20 20 20 20 20 20 20 0a  |               .|
00000080  01 00 00 00 00 00 00 00  02 00 00 00 00 00 00 00  |................|
00000090  03 00 00 00 00 00 00 00  04 00 00 00 00 00 00 00  |................|
000000a0  05 00 00 00 00 00 00 00  06 00 00 00 00 00 00 00  |................|
```

### CODE REVIEW

File: TensorLib/NumpyRepr.lean [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+  let id <- quoted parseToken
+  colon
+  if id == "descr" then
+    let v <- quoted parseToken
+    let d <- NumpyDtype.fromString v
+    modify (fun s => { s with descr := some d })
+  else if id == "fortran_order" then
+    let v <- parseToken
+    let s <- get
+    match v with
```
Comment (govereau) [inline TensorLib/NumpyRepr.lean:237]: This is where I though you may use the bool fromString?

**Review by govereau — APPROVED** [review]
Looks great!

File: TensorLib/NumpyRepr.lean [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+| "<u4" => .ok uint32
+| "<u8" => .ok uint64
+| "<f2" => .ok float16
+| "<f4" => .ok float32
+| "<f8" => .ok float64
+| _ => .error s!"Can't parse {s} as a dtype"
+
+end NumpyDtype
+
+def NumpyBool.fromString (s: String): Option Bool := match s with
```
[the remark this answers is not in the document]
Comment (seanmcl) [reply TensorLib/NumpyRepr.lean:56]: Lol. Yes, this is an oversight due to me getting bogged down in the parser and forgetting to add it in. Will make the types nicer! Will do and will add some tests with different dtypes.

File: TensorLib/NumpyRepr.lean [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+variable {T : Type}
+
+def debug (msg : String) : PState Unit := do
+  modify (fun s => { s with debug := msg :: s.debug })
+
+def whitespace : PState Unit := do
+  let s <- get
+  let mut count := 0
+  for i in [s.index : s.headerEnd] do
+    if s.source[i]! == ' '.toUInt8 then
```
[the remark this answers is not in the document]
Comment (seanmcl) [reply TensorLib/NumpyRepr.lean:136]: Yeah... looks weird and it's underspecified in the docs. They are clear that 0x20 is the padding but the space in the python object syntax just appears to be the same, but indeed could be something else. Maybe I'll convert to Char and use `isWhitespace`.

File: TensorLib/NumpyRepr.lean (hunk 1 above)
Comment (seanmcl) [reply TensorLib/NumpyRepr.lean:237]: Yep.

File: Main.lean [hunk 4]
```diff
@@ -1,4 +1,39 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: John Tristan, Paul Govereau, Sean McLaughlin
```
Comment (seanmcl) [inline Main.lean:4]: Fixing this to Jean-Baptiste.

**Pull request merged**